File: ios/EndlessRumination/Models/Lens.swift
import SwiftUI

struct Lens: Identifiable {
    let id: Int
    let name: String
    let emoji: String
    let color: Color
    let bgColor: Color

    var index: Int { id }

    /// Free-tier users see the first 5 lenses only
    static let freeLensCount = 5

    /// These lens indices use Sonnet ("Wise") even for free-tier users
    static let wiseLensIndices: Set<Int> = [1, 9]

    /// Whether this lens is available to free-tier users
    var isFree: Bool { id < Self.freeLensCount }

    /// Whether this lens is powered by Sonnet (shown as "Wise")
    var isWise: Bool { Self.wiseLensIndices.contains(id) }
}

extension Lens {
    static let all: [Lens] = [
        Lens(id: 0,  name: "The Comedian",        emoji: "😂",  color: Color(hex: 0xFF6B9D), bgColor: Color(hex: 0xFF6B9D).opacity(0.15)),
        Lens(id: 1,  name: "The Stoic",            emoji: "🏛",  color: Color(hex: 0xC9A84C), bgColor: Color(hex: 0xC9A84C).opacity(0.15)),
        Lens(id: 2,  name: "The Nihilist",          emoji: "🕳",  color: Color(hex: 0x8A8690), bgColor: Color.white.opacity(0.06)),
        Lens(id: 3,  name: "The Optimist",          emoji: "☀️",  color: Color(hex: 0x3ECF8E), bgColor: Color(hex: 0x3ECF8E).opacity(0.15)),
        Lens(id: 4,  name: "The Pessimist",         emoji: "⛈",   color: Color(hex: 0xFF4757), bgColor: Color(hex: 0xFF4757).opacity(0.15)),
        Lens(id: 5,  name: "Your Best Friend",      emoji: "🫂",  color: Color(hex: 0x4A7CFF), bgColor: Color(hex: 0x4A7CFF).opacity(0.15)),
        Lens(id: 6,  name: "The Poet",              emoji: "🪶",  color: Color(hex: 0x9B6DFF), bgColor: Color(hex: 0x9B6DFF).opacity(0.15)),
        Lens(id: 7,  name: "A Five-Year-Old",       emoji: "🧸",  color: Color(hex: 0xF0C832), bgColor: Color(hex: 0xF0C832).opacity(0.15)),
        Lens(id: 8,  name: "The CEO",               emoji: "📊",  color: Color(hex: 0xF0ECE4), bgColor: Color(hex: 0xF0ECE4).opacity(0.08)),
        Lens(id: 9,  name: "The Therapist",         emoji: "🪷",  color: Color(hex: 0x00D4AA), bgColor: Color(hex: 0x00D4AA).opacity(0.15)),
        Lens(id: 10, name: "Your Grandma",          emoji: "🍪",  color: Color(hex: 0xE8653A), bgColor: Color(hex: 0xE8653A).opacity(0.15)),
        Lens(id: 11, name: "The Alien",             emoji: "👽",  color: Color(hex: 0x4AFFB4), bgColor: Color(hex: 0x4AFFB4).opacity(0.12)),
        Lens(id: 12, name: "The Historian",          emoji: "📜",  color: Color(hex: 0xD4A843), bgColor: Color(hex: 0xD4A843).opacity(0.12)),
        Lens(id: 13, name: "The Philosopher",        emoji: "🦉",  color: Color(hex: 0xB08AFF), bgColor: Color(hex: 0xB08AFF).opacity(0.12)),
        Lens(id: 14, name: "Future You",             emoji: "⌛",   color: Color(hex: 0x6E9FFF), bgColor: Color(hex: 0x6E9FFF).opacity(0.12)),
        Lens(id: 15, name: "Drill Sergeant",         emoji: "🎖",  color: Color(hex: 0xC8C0B4), bgColor: Color(hex: 0xC8C0B4).opacity(0.10)),
        Lens(id: 16, name: "The Monk",               emoji: "🧘",  color: Color(hex: 0x40DFB0), bgColor: Color(hex: 0x40DFB0).opacity(0.10)),
        Lens(id: 17, name: "The Scientist",          emoji: "🔬",  color: Color(hex: 0x5A8CFF), bgColor: Color(hex: 0x5A8CFF).opacity(0.12)),
        Lens(id: 18, name: "Conspiracy Theorist",    emoji: "🔺",  color: Color(hex: 0xE8B830), bgColor: Color(hex: 0xE8B830).opacity(0.12)),
        Lens(id: 19, name: "Your Dog",               emoji: "🐕",  color: Color(hex: 0xF0A070), bgColor: Color(hex: 0xF0A070).opacity(0.12)),
    ]

    static let freeLenses: [Lens] = Array(all.prefix(freeLensCount))

    static func lens(at index: Int) -> Lens {
        all[index % all.count]
    }

    /// Unified display info for both base lenses (0-19) and pack voices (20-39).
    static func displayInfo(at index: Int) -> (name: String, emoji: String, color: Color, bgColor: Color) {
        if index < all.count {
            let l = all[index]
            return (l.name, l.emoji, l.color, l.bgColor)
        }
        if let v = VoicePack.voice(at: index) {
            return (v.name, v.emoji, v.color, v.bgColor)
        }
        // Fallback
        return ("Unknown", "?", ERColors.dimText, ERColors.inputBackground)
    }
}

File: android/app/src/main/kotlin/com/endlessrumination/model/Lens.kt
package com.endlessrumination.model

import androidx.compose.ui.graphics.Color

data class Lens(
    val id: Int,
    val name: String,
    val emoji: String,
    val color: Color,
    val bgColor: Color
) {
    /** Whether this lens is available to free-tier users */
    val isFree: Boolean get() = id < FREE_LENS_COUNT

    /** Whether this lens is powered by Sonnet (shown as "Wise") */
    val isWise: Boolean get() = wiseLensIndices.contains(id)

    companion object {
        const val FREE_LENS_COUNT = 5

        /** These lens indices use Sonnet ("Wise") even for free-tier users */
        val wiseLensIndices: Set<Int> = setOf(1, 9)

        val freeLenses: List<Lens> by lazy { all.take(FREE_LENS_COUNT) }

        val all: List<Lens> = listOf(
            Lens(0,  "The Comedian",       "\uD83D\uDE02", Color(0xFFFF6B9D), Color(0xFFFF6B9D).copy(alpha = 0.15f)),
            Lens(1,  "The Stoic",          "\uD83C\uDFDB", Color(0xFFC9A84C), Color(0xFFC9A84C).copy(alpha = 0.15f)),
            Lens(2,  "The Nihilist",       "\uD83D\uDD73", Color(0xFF8A8690), Color.White.copy(alpha = 0.06f)),
            Lens(3,  "The Optimist",       "\u2600\uFE0F", Color(0xFF3ECF8E), Color(0xFF3ECF8E).copy(alpha = 0.15f)),
            Lens(4,  "The Pessimist",      "\u26C8",       Color(0xFFFF4757), Color(0xFFFF4757).copy(alpha = 0.15f)),
            Lens(5,  "Your Best Friend",   "\uD83E\uDEC2", Color(0xFF4A7CFF), Color(0xFF4A7CFF).copy(alpha = 0.15f)),
            Lens(6,  "The Poet",           "\uD83E\uDEB6", Color(0xFF9B6DFF), Color(0xFF9B6DFF).copy(alpha = 0.15f)),
            Lens(7,  "A Five-Year-Old",    "\uD83E\uDDF8", Color(0xFFF0C832), Color(0xFFF0C832).copy(alpha = 0.15f)),
            Lens(8,  "The CEO",            "\uD83D\uDCCA", Color(0xFFF0ECE4), Color(0xFFF0ECE4).copy(alpha = 0.08f)),
            Lens(9,  "The Therapist",      "\uD83E\uDEB7", Color(0xFF00D4AA), Color(0xFF00D4AA).copy(alpha = 0.15f)),
            Lens(10, "Your Grandma",       "\uD83C\uDF6A", Color(0xFFE8653A), Color(0xFFE8653A).copy(alpha = 0.15f)),
            Lens(11, "The Alien",          "\uD83D\uDC7D", Color(0xFF4AFFB4), Color(0xFF4AFFB4).copy(alpha = 0.12f)),
            Lens(12, "The Historian",      "\uD83D\uDCDC", Color(0xFFD4A843), Color(0xFFD4A843).copy(alpha = 0.12f)),
            Lens(13, "The Philosopher",    "\uD83E\uDD89", Color(0xFFB08AFF), Color(0xFFB08AFF).copy(alpha = 0.12f)),
            Lens(14, "Future You",         "\u231B",       Color(0xFF6E9FFF), Color(0xFF6E9FFF).copy(alpha = 0.12f)),
            Lens(15, "Drill Sergeant",     "\uD83C\uDF96", Color(0xFFC8C0B4), Color(0xFFC8C0B4).copy(alpha = 0.10f)),
            Lens(16, "The Monk",           "\uD83E\uDDD8", Color(0xFF40DFB0), Color(0xFF40DFB0).copy(alpha = 0.10f)),
            Lens(17, "The Scientist",      "\uD83D\uDD2C", Color(0xFF5A8CFF), Color(0xFF5A8CFF).copy(alpha = 0.12f)),
            Lens(18, "Conspiracy Theorist","\uD83D\uDD3A", Color(0xFFE8B830), Color(0xFFE8B830).copy(alpha = 0.12f)),
            Lens(19, "Your Dog",           "\uD83D\uDC15", Color(0xFFF0A070), Color(0xFFF0A070).copy(alpha = 0.12f)),
        )

        /** Unified display info for both base lenses (0-19) and pack voices (20-39). */
        fun displayInfo(index: Int): DisplayInfo {
            if (index < all.size) {
                val l = all[index]
                return DisplayInfo(l.name, l.emoji, l.color, l.bgColor)
            }
            val v = VoicePack.voiceAt(index)
            if (v != null) {
                return DisplayInfo(v.name, v.emoji, v.color, v.bgColor)
            }
            return DisplayInfo("Unknown", "?", Color(0xFF4A4650), Color(0xFF1A1A20))
        }
    }

    data class DisplayInfo(
        val name: String,
        val emoji: String,
        val color: Color,
        val bgColor: Color
    )
}
